sklearn/model_selection/_search.py

```diff
@@ -558,53 +588,226 @@ def _fit(self, X, y, labels, parameter_iterable):
           for train, test in cv.split(X, y, labels))
 
         # Out is a list of triplet: score, estimator, n_test_samples
-        n_fits = len(out)
+        self.n_fits_ = len(out)
 
-        scores = list()
-        grid_scores = list()
-        for grid_start in range(0, n_fits, n_splits):
-            n_test_samples = 0
-            score = 0
-            all_scores = []
-            for this_score, this_n_test_samples, _, parameters in \
+        self.n_candidates_ = self.n_fits_ // n_splits
+        self.n_parameters_ = len(self.parameter_names_)
+
+        res_shape = (self.n_candidates_,)
+
+        search_results = dict()
+
+        # Lets not initite this everytime and reuse the same array.
+        all_scores = np.empty((n_splits,), dtype=np.float64)
+
+        metric = self.metric_name_
+
+        for param in self.parameter_names_:
+            # One column to record the values of each parameter
+            search_results["param_%s" % param] = (
+                np.ma.masked_all(res_shape, dtype=object))
+
+        # Make a column for each split of each metric
+        for split_i in range(n_splits):
+            search_results["%s_split_%s" % (metric, split_i)] = (
+                np.empty(res_shape, dtype=np.float64))
+        search_results["%s_mean" % metric] = np.empty(res_shape,
+                                                      dtype=np.float64)
+        search_results["%s_rank" % metric] = np.empty(res_shape, dtype=int)
+
+        for fit_i, grid_start in enumerate(range(0, self.n_fits_, n_splits)):
+            n_test_samples_total = 0
+            mean_score = 0
+
+            split_i = -1
+            for score_i, n_test_samples_i, _, parameters in \
                     out[grid_start:grid_start + n_splits]:
-                all_scores.append(this_score)
+                split_i += 1
+                # Record the score/n_test_samples for the i-th split
+                # of the current parameter setting candidate.
+                all_scores[split_i] = score_i
+
                 if self.iid:
-                    this_score *= this_n_test_samples
-                    n_test_samples += this_n_test_samples
-                score += this_score
+                    score_i *= n_test_samples_i
+                    n_test_samples_total += n_test_samples_i
+
+                mean_score += score_i
+                search_results["%s_split_%s" %
+                               (metric, split_i)][fit_i] = score_i
+
             if self.iid:
-                score /= float(n_test_samples)
+                mean_score = all_scores.sum() / float(n_test_samples_total)
```

Comment:
Can we please avoid `float` casting and instead ensure we have `from __future__ import division` up top?